README.markdown
Resque
======

[![Gem Version](https://badge.fury.io/rb/resque.svg)](https://rubygems.org/gems/resque)
[![Build Status](https://github.com/resque/resque/actions/workflows/ci.yml/badge.svg)](https://github.com/resque/resque/actions/workflows/ci.yml)

Introduction
------------

Resque (pronounced like "rescue") is a Redis-backed library for creating
background jobs, placing those jobs on multiple queues, and processing
them later.

For the backstory, philosophy, and history of Resque's beginnings, please see [the blog post](http://github.com/blog/542-introducing-resque) (2009).

Background jobs can be any Ruby class or module that responds to
`perform`. Your existing classes can easily be converted to background
jobs or you can create new classes specifically to do work. Or, you
can do both.

Resque is heavily inspired by DelayedJob (which rocks) and comprises
three parts:

1. A Ruby library for creating, querying, and processing jobs
2. A Rake task for starting a worker which processes jobs
3. A Sinatra app for monitoring queues, jobs, and workers.

Resque workers can be  given multiple queues (a "queue list"),
distributed between multiple machines,
run anywhere with network access to the Redis server,
support priorities, are resilient to memory bloat / "leaks,"
tell you what they're doing, and expect failure.

Resque queues are persistent; support constant time, atomic push and
pop (thanks to Redis); provide visibility into their contents; and
store jobs as simple JSON packages.

The Resque frontend tells you what workers are doing, what workers are
not doing, what queues you're using, what's in those queues, provides
general usage stats, and helps you track failures.

Resque now supports Ruby 2.3.0 and above.
We will also only be supporting Redis 3.0 and above going forward.

### Note on the future of Resque

Would you like to be involved in Resque? Do you have thoughts about what
Resque should be and do going forward? There's currently an [open discussion here](https://github.com/resque/resque/issues/1759)
on just that topic, so please feel free to join in. We'd love to hear your thoughts
and/or have people volunteer to be a part of the project!

Example
-------

Resque jobs are Ruby classes (or modules) which respond to the
`perform` method. Here's an example:


``` ruby
class Archive
  @queue = :file_serve

  def self.perform(repo_id, branch = 'master')
    repo = Repository.find(repo_id)
    repo.create_archive(branch)
  end
end
```

The `@queue` class instance variable determines which queue `Archive`
jobs will be placed in. Queues are arbitrary and created on the fly -
you can name them whatever you want and have as many as you want.

To place an `Archive` job on the `file_serve` queue, we might add this
to our application's pre-existing `Repository` class:

``` ruby
class Repository
  def async_create_archive(branch)
    ResqueSqs.enqueue(Archive, self.id, branch)
  end
end
```

Now when we call `repo.async_create_archive('masterbrew')` in our
application, a job will be created and placed on the `file_serve`
queue.

Later, a worker will run something like this code to process the job:

``` ruby
klass, args = ResqueSqs.reserve(:file_serve)
klass.perform(*args) if klass.respond_to? :perform
```

Which translates to:

``` ruby
Archive.perform(44, 'masterbrew')
```

Let's start a worker to run `file_serve` jobs:

    $ cd app_root
    $ QUEUE=file_serve rake resque_sqs:work

This starts one Resque worker and tells it to work off the
`file_serve` queue. As soon as it's ready it'll try to run the
`ResqueSqs.reserve` code snippet above and process jobs until it can't
find any more, at which point it will sleep for a small period and
repeatedly poll the queue for more jobs.

Installation
------------

Add the gem to your Gemfile:

    gem 'resque_sqs'

Next, install it with Bundler:

    $ bundle

#### Rack

In your Rakefile, or some other file in `lib/tasks` (ex: `lib/tasks/resque.rake`), load the resque rake tasks:

``` ruby
require 'resque_sqs'
require 'resque_sqs/tasks'
require 'your/app' # Include this line if you want your workers to have access to your application
```

#### Rails

To make resque specific changes, you can override the `resque_sqs:setup` job in `lib/tasks` (ex: `lib/tasks/resque.rake`). GitHub's setup task looks like this:

``` ruby
task "resque_sqs:setup" => :environment do
  Grit::Git.git_timeout = 10.minutes
end
```

We don't want the `git_timeout` as high as 10 minutes in our web app,
but in the Resque workers it's fine.

Running Workers
---------------

Resque workers are rake tasks that run forever. They basically do this:

``` ruby
start
loop do
  if job = reserve
    job.process
  else
    sleep 5 # Polling frequency = 5
  end
end
shutdown
```

Starting a worker is simple:

    $ QUEUE=* rake resque_sqs:work

Or, you can start multiple workers:

    $ COUNT=2 QUEUE=* rake resque_sqs:workers

This will spawn two Resque workers, each in its own process. Hitting
ctrl-c should be sufficient to stop them all.

#### Priorities and Queue Lists

Resque doesn't support numeric priorities but instead uses the order
of queues you give it. We call this list of queues the "queue list."

Let's say we add a `warm_cache` queue in addition to our `file_serve`
queue. We'd now start a worker like so:

    $ QUEUES=file_serve,warm_cache rake resque_sqs:work

When the worker looks for new jobs, it will first check
`file_serve`. If it finds a job, it'll process it then check
`file_serve` again. It will keep checking `file_serve` until no more
jobs are available. At that point, it will check `warm_cache`. If it
finds a job it'll process it then check `file_serve` (repeating the
whole process).

In this way you can prioritize certain queues. At GitHub we start our
workers with something like this:

    $ QUEUES=critical,archive,high,low rake resque_sqs:work

Notice the `archive` queue - it is specialized and in our future
architecture will only be run from a single machine.

At that point we'll start workers on our generalized background
machines with this command:

    $ QUEUES=critical,high,low rake resque_sqs:work

And workers on our specialized archive machine with this command:

    $ QUEUE=archive rake resque_sqs:work

#### Running All Queues

If you want your workers to work off of every queue, including new
queues created on the fly, you can use a splat:

    $ QUEUE=* rake resque_sqs:work

Queues will be processed in alphabetical order.

Or, prioritize some queues above `*`:

    # QUEUE=critical,* rake resque_sqs:work

#### Running All Queues Except for Some

If you want your workers to work off of all queues except for some,
you can use negation:

    $ QUEUE=*,!low rake resque_sqs:work

Negated globs also work. The following will instruct workers to work
off of all queues except those beginning with `file_`:

    $ QUEUE=*,!file_* rake resque_sqs:work

Note that the order in which negated queues are specified does not
matter, so `QUEUE=*,!file_*` and `QUEUE=!file_*,*` will have the same
effect.

#### Process IDs (PIDs)

There are scenarios where it's helpful to record the PID of a resque
worker process.  Use the PIDFILE option for easy access to the PID:

    $ PIDFILE=./resque.pid QUEUE=file_serve rake resque_sqs:work

#### Running in the background

There are scenarios where it's helpful for
the resque worker to run itself in the background (usually in combination with
PIDFILE).  Use the BACKGROUND option so that rake will return as soon as the
worker is started.

    $ PIDFILE=./resque.pid BACKGROUND=yes QUEUE=file_serve rake resque_sqs:work

#### Polling frequency

You can pass an INTERVAL option which is a float representing the polling frequency.
The default is 5 seconds, but for a semi-active app you may want to use a smaller value.

    $ INTERVAL=0.1 QUEUE=file_serve rake resque_sqs:work

The Front End
-------------

Resque comes with a Sinatra-based front end for seeing what's up with
your queue.

![The Front End](https://camo.githubusercontent.com/64d150a243987ffbc33f588bd6d7722a0bb8d69a/687474703a2f2f7475746f7269616c732e6a756d7073746172746c61622e636f6d2f696d616765732f7265737175655f6f766572766965772e706e67)

#### Standalone

If you've installed Resque as a gem running the front end standalone is easy:

    $ resque-web

It's a thin layer around `rackup` so it's configurable as well:

    $ resque-web -p 8282

If you have a Resque config file you want evaluated just pass it to
the script as the final argument:

    $ resque-web -p 8282 rails_root/config/initializers/resque.rb

You can also set the namespace directly using `resque-web`:

    $ resque-web -p 8282 -N myapp

or set the Redis connection string if you need to do something like select a different database:

    $ resque-web -p 8282 -r localhost:6379:2

#### Passenger

Using Passenger? Resque ships with a `config.ru` you can use. See
Phusion's guide:

Apache: <https://www.phusionpassenger.com/library/deploy/apache/deploy/ruby/>
Nginx: <https://www.phusionpassenger.com/library/deploy/nginx/deploy/ruby/>

#### Rack::URLMap

If you want to load Resque on a subpath, possibly alongside other
apps, it's easy to do with Rack's `URLMap`:

``` ruby
require 'resque_sqs/server'

run Rack::URLMap.new \
  "/"       => Your::App.new,
  "/resque" => ResqueSqs::Server.new
```

Check `examples/demo/config.ru` for a functional example (including
HTTP basic auth).

#### Rails

You can also mount Resque on a subpath in your existing Rails app by adding `require 'resque_sqs/server'` to the top of your routes file or in an initializer then adding this to `routes.rb`:

``` ruby
mount ResqueSqs::Server.new, :at => "/resque"
```

Jobs
----

What should you run in the background? Anything that takes any time at
all. Slow INSERT statements, disk manipulating, data processing, etc.

At GitHub we use Resque to process the following types of jobs:

* Warming caches
* Counting disk usage
* Building tarballs
* Building Rubygems
* Firing off web hooks
* Creating events in the db and pre-caching them
* Building graphs
* Deleting users
* Updating our search index

As of writing we have about 35 different types of background jobs.

Keep in mind that you don't need a web app to use Resque - we just
mention "foreground" and "background" because they make conceptual
sense. You could easily be spidering sites and sticking data which
needs to be crunched later into a queue.

#### Persistence

Jobs are persisted to queues as JSON objects. Let's take our `Archive`
example from above. We'll run the following code to create a job:

``` ruby
repo = Repository.find(44)
repo.async_create_archive('masterbrew')
```

The following JSON will be stored in the `file_serve` queue:

``` javascript
{
    'class': 'Archive',
    'args': [ 44, 'masterbrew' ]
}
```

Because of this your jobs must only accept arguments that can be JSON encoded.

So instead of doing this:

``` ruby
ResqueSqs.enqueue(Archive, self, branch)
```

do this:

``` ruby
ResqueSqs.enqueue(Archive, self.id, branch)
```

This is why our above example (and all the examples in `examples/`)
uses object IDs instead of passing around the objects.

While this is less convenient than just sticking a marshaled object
in the database, it gives you a slight advantage: your jobs will be
run against the most recent version of an object because they need to
pull from the DB or cache.

If your jobs were run against marshaled objects, they could
potentially be operating on a stale record with out-of-date information.


#### send_later / async

Want something like DelayedJob's `send_later` or the ability to use
instance methods instead of just methods for jobs? See the `examples/`
directory for goodies.

We plan to provide first class `async` support in a future release.


#### Failure

If a job raises an exception, it is logged and handed off to the
`ResqueSqs::Failure` module. Failures are logged either locally in Redis
or using some different backend. To see exceptions while developing,
see details below under Logging.

For example, Resque ships with Airbrake support. To configure it, put
the following into an initialisation file or into your rake job:

``` ruby
# send errors which occur in background jobs to redis and airbrake
require 'resque_sqs/failure/multiple'
require 'resque_sqs/failure/redis'
require 'resque_sqs/failure/airbrake'

ResqueSqs::Failure::Multiple.classes = [ResqueSqs::Failure::Redis, ResqueSqs::Failure::Airbrake]
ResqueSqs::Failure.backend = ResqueSqs::Failure::Multiple
```

Keep this in mind when writing your jobs: you may want to throw
exceptions you would not normally throw in order to assist debugging.


#### Rails example

If you are using ActiveJob here's how your job definition will look:

``` ruby
class ArchiveJob < ApplicationJob
  queue_as :file_serve

  def perform(repo_id, branch = 'master')
    repo = Repository.find(repo_id)
    repo.create_archive(branch)
  end
end
```

``` ruby
class Repository
  def async_create_archive(branch)
    ArchiveJob.perform_later(self.id, branch)
  end
end
```

It is important to run `ArchiveJob.perform_later(self.id, branch)` rather than `ResqueSqs.enqueue(Archive, self.id, branch)`.
Otherwise Resque will process the job without actually doing anything.
Even if you put an obviously buggy line like `0/0` in the `perform` method,
the job will still succeed.


Configuration
-------------

#### Redis

You may want to change the Redis host and port Resque connects to, or
set various other options at startup.

Resque has a `redis` setter which can be given a string or a Redis
object. This means if you're already using Redis in your app, Resque
can re-use the existing connection.

String: `ResqueSqs.redis = 'localhost:6379'`

Redis: `ResqueSqs.redis = $redis`

For our rails app we have a `config/initializers/resque.rb` file where
we load `config/resque.yml` by hand and set the Redis information
appropriately.

Here's our `config/resque.yml`:

    development: localhost:6379
    test: localhost:6379
    staging: redis1.se.github.com:6379
    fi: localhost:6379
    production: <%= ENV['REDIS_URL'] %>

And our initializer:

``` ruby
rails_root = ENV['RAILS_ROOT'] || File.dirname(__FILE__) + '/../..'
rails_env = ENV['RAILS_ENV'] || 'development'
config_file = rails_root + '/config/resque.yml'

resque_config = YAML::load(ERB.new(IO.read(config_file)).result)
ResqueSqs.redis = resque_config[rails_env]
```

Easy peasy! Why not just use `RAILS_ROOT` and `RAILS_ENV`? Because

HISTORY.md
## 2.5.0

### Fixed

* Replace `File.exists?` with `File.exist?` (#1846)
* Escape `ResqueSqs.redis_id` for stats page (#1834)
* Escape resque info values (#1832)
* Correctly show the values of hash and none type on stats tab (#1838)
* Fix logging the worker name when starting the task (#1837)

### Added

* Raise an error when no available Rack server was found (#1836)
* Move code in `ResqueSqs::Server.helpers` block into a module to make it testable (#1851)

## 2.4.0

### Fixed

* Remove `thread_safe` arg in Redis instantiaons (#1822)
* Updated Test Matrix for Redis 5.0 (#1824)
* Fix redis-rb 4.8 deprecation warnings (#1827)
* Fix redis-rb 5.0 compatibility issues (#1828)

## 2.3.0

### Fixed

* Fix NoMethodError: undefined method 'application' for Rails:Module when Rails module is defined but not a full Rails app (#1799)
* Fix deprecation warnings in Redis#pipelined for redis 4.6 (#1806)
* Add Ruby syntax highlighting to remaining markdown docs (#1802)
* Fix pagination section (#1809)
* Fix `before_run` undefined local variable or method `runner` (#1811)

### Added

* Added support for pausing all workers by setting the Redis key `pause-all-workers` to string value "true" (#1803)

## 2.2.1

### Fixed

* Escape HTML from the params to avoid XSS (#1790)
* Remove vegas as a dependency (#1780)

### Added

* Add support for specifying queues that workers should ignore when using globs
* Allow plugins to have spaces in their name (#1718)

## 2.2.0

### Added

* Add 'Retry Failed Jobs' button to Failed page

### Fixed

* Loading railtie when it did not exist #1765
* Fix closing parent Redis connection in child process after fork
* Fix the failed queues list on /queues UI when the RedisMultiQueue backend is used #1638

## 2.1.0

### Security

* Fix XSS via URL path in admin web UI queues view #1687
* Replace onclick handlers in server code to support Content Security Policies that don't allow 'unsafe-inline'
* Update jQuery from 1.12.4 to 3.6.0

### Added

* Add requeue_queue method to ResqueSqs::Failure::Multiple #1659
* Confirmation prompt in admin front-end before submitting the retry of all failed jobs. #1753
* Railtie for default rake task setup when in Rails. #1715
* Added two new hooks.
  - `queue_empty` when the job queue empties and the worker becomes idle
  - `worker_exit` when the worker exits

  See [docs/HOOKS.md](http://github.com/resque/resque/blob/master/docs/HOOKS.md) for
  further details. (@jeremywadsack)

### Fixed

* live poller shouldn't restart itself until it successds or fails. #1740
* Fix parsing worker_id when queue name includes colon. #1691
* Prune workers which haven't been registered but have set a heartbeat. #1751
* `ResqueSqs::Failure::Multiple.remove` did not pass on the queue parameter

## 2.0.0 (2018-11-06)

### Fixed
* Fix Airbrake failure backend
* Fix failed jobs page "argument out of range" error

### Changed
* Remove support for Rubies < 2.3
* Remove support to Rails < 4
* Reduce the number of redis calls when trying to get the list of queues
* Only run `eager_load!` if `Rails.application.config.eager_load` is true
* Don't display log message if running without hooks
* Add Support to Redis 4.0
* Drop complex Redis identifier logic in favor of simple inspect
* When a child process is killed, report the signal it was terminated with
* Report a job that pruned worker was processing

### Added

* Allow to configure statistic data store

## 1.27.4 (2017-04-15)

### Fixed
* Fix issue where removing a failure from Resque web didn't work when using `RedisMultiQueue` backend.

## 1.27.3 (2017-04-10)

### Fixed
* Fix issue where initializing a data store would attempt to hit Redis, even when ResqueSqs.inline = true

## 1.27.2 (2017-02-20)

### Fixed
* Require "redis/distributed" in worker.rb to allow proper rescuing
* Fallback to server time if Redis time won't work (restores Redis 2.4 support)

## 1.27.1 (2017-02-13)

### Fixed
* Show actual jobs names in web view using ActiveJob (@martnst)

## 1.27.0 (2017-02-08)

### Fixed
* Fix issue where calling Worker.find, Worker.all, or Worker.working from withing
  a running job would rewrite the PID file with the PID of the forked worker.
  This causes a change to the Worker#new API that may affect plugin
  implementations. See Worker#new and Worker#prepare for details. (@jeremywadsack)
* Workers queried out now have the correct hostname (@crazymykl)
* Fix race condition on worker startup (@stevedomin)
* No longer triggers verbose logging if env variables are not set (@ldnunes)
* resque/failed/requeue/all when using Redis::Failure::Multiple no longer raises an exception (@ale7714)
* Improve forking to avoid having a child process escape its code (@dylanahsmith)
* Workers now use server time rather than their own time to maintain heartbeats (@fw42)
* Run eager load hooks for Rails applications versioned 4.x and up (@abhi-patel)
* Fix bug when encountering an error while pruning workers (Joakim Kolsjö and Tomas Skogberg)
* Children write to PIDFILE immediately after forking, fixing issues when reconnecting to Redis is slow (@fimmtiu)

### Changed
* Update jQuery from 1.3.2 to 1.12.4 (@chrisccerami)
* No longer user Thread.kill to stop heartbeat (@Sinjo)

### Added
* Resque Web UI now prompts for confirmation on clearing failed jobs (Markus Olsen)
* Adds process status to DirtyExit exception when job is killed via signal (@MishaConway)

## 1.26.0 (2016-03-10)

This changelog is a bit incomplete. We will be much stricter about the changelog for
the next release.

* rake: when BACKGROUND, ensure worker.reconnect after daemonizing (@yaauie)
* worker: when searching workers, Worker#pid properly reflects pid of the found
  worker instead of the current pid. (@yaauie)
* Add support for RESQUE_PROCLINE_PREFIX environment variable to prefix
  procline strings with a static identifier. (@rbroemeling)
* ResqueSqs::Worker logs errors at the correct logging level (@chrisccerami)

## 1.25.2 (2014-3-3)


lib/resque_sqs.rb
require 'mono_logger'
require 'redis/namespace'
require 'forwardable'

require 'resque_sqs/version'

require 'resque_sqs/errors'

require 'resque_sqs/failure'
require 'resque_sqs/failure/base'

require 'resque_sqs/helpers'
require 'resque_sqs/stat'
require 'resque_sqs/logging'
require 'resque_sqs/log_formatters/quiet_formatter'
require 'resque_sqs/log_formatters/verbose_formatter'
require 'resque_sqs/log_formatters/very_verbose_formatter'
require 'resque_sqs/job'
require 'resque_sqs/worker'
require 'resque_sqs/plugin'
require 'resque_sqs/data_store'
require 'resque_sqs/thread_signal'

require 'resque_sqs/vendor/utf8_util'

require 'resque_sqs/railtie' if defined?(Rails::Railtie)
require 'aws-sdk-sqs/client'

module ResqueSqs
  include Helpers
  extend self

  # Given a Ruby object, returns a string suitable for storage in a
  # queue.
  def encode(object)
    if MultiJson.respond_to?(:dump) && MultiJson.respond_to?(:load)
      MultiJson.dump object
    else
      MultiJson.encode object
    end
  end

  # Given a string, returns a Ruby object.
  def decode(object)
    return unless object

    begin
      if MultiJson.respond_to?(:dump) && MultiJson.respond_to?(:load)
        MultiJson.load object
      else
        MultiJson.decode object
      end
    rescue ::MultiJson::DecodeError => e
      raise Helpers::DecodeException, e.message, e.backtrace
    end
  end

  # Given a word with dashes, returns a camel cased version of it.
  #
  # classify('job-name') # => 'JobName'
  def classify(dashed_word)
    dashed_word.split('-').map(&:capitalize).join
  end

  # Tries to find a constant with the name specified in the argument string:
  #
  # constantize("Module") # => Module
  # constantize("Test::Unit") # => Test::Unit
  #
  # The name is assumed to be the one of a top-level constant, no matter
  # whether it starts with "::" or not. No lexical context is taken into
  # account:
  #
  # C = 'outside'
  # module M
  #   C = 'inside'
  #   C # => 'inside'
  #   constantize("C") # => 'outside', same as ::C
  # end
  #
  # NameError is raised when the constant is unknown.
  def constantize(camel_cased_word)
    camel_cased_word = camel_cased_word.to_s

    if camel_cased_word.include?('-')
      camel_cased_word = classify(camel_cased_word)
    end

    names = camel_cased_word.split('::')
    names.shift if names.empty? || names.first.empty?

    constant = Object
    names.each do |name|
      args = Module.method(:const_get).arity != 1 ? [false] : []

      if constant.const_defined?(name, *args)
        constant = constant.const_get(name)
      else
        constant = constant.const_missing(name)
      end
    end
    constant
  end

  extend ::Forwardable

  # Accepts:
  #   1. A 'hostname:port' String
  #   2. A 'hostname:port:db' String (to select the Redis db)
  #   3. A 'hostname:port/namespace' String (to set the Redis namespace)
  #   4. A Redis URL String 'redis://host:port'
  #   5. An instance of `Redis`, `Redis::Client`, `Redis::DistRedis`,
  #      or `Redis::Namespace`.
  #   6. An Hash of a redis connection {:host => 'localhost', :port => 6379, :db => 0}
  def data_store=(data_store)
    raise 'ResqueSqs::DataStore must be provided' unless data_store.is_a?(ResqueSqs::DataStore)

    @data_store = data_store
  end

  def data_store
    @data_store
  end

  def redis
    return @data_store.redis if @data_store

    redis_connection = Redis.respond_to?(:connect) ? Redis.connect : "localhost:6379"
    @data_store = ResqueSqs::DataStore.new(redis_connection, Aws::SQS::Client.new)
    @data_store.redis
  end

  def redis_id
    @data_store.identifier
  end

  # Set the data store for the processed and failed statistics.
  #
  # By default it uses the same as `ResqueSqs.redis`, but different stores can be used.
  #
  # A custom store needs to obey the following API to work correctly
  #
  # class NullDataStore
  #   # Returns the current value for the given stat.
  #   def stat(stat)
  #   end
  #
  #   # Increments the stat by the given value.
  #   def increment_stat(stat, by)
  #   end
  #
  #   # Decrements the stat by the given value.
  #   def decrement_stat(stat, by)
  #   end
  #
  #   # Clear the values for the given stat.
  #   def clear_stat(stat)
  #   end
  # end
  def stat_data_store=(stat_data_store)
    ResqueSqs::Stat.data_store = stat_data_store
  end

  # Returns the data store for the statistics module.
  def stat_data_store
    ResqueSqs::Stat.data_store
  end

  # Set or retrieve the current logger object
  attr_accessor :logger

  DEFAULT_HEARTBEAT_INTERVAL = 60
  DEFAULT_PRUNE_INTERVAL = DEFAULT_HEARTBEAT_INTERVAL * 5

  # Defines how often a Resque worker updates the heartbeat key. Must be less
  # than the prune interval.
  attr_writer :heartbeat_interval
  def heartbeat_interval
    if defined? @heartbeat_interval
      @heartbeat_interval
    else
      DEFAULT_HEARTBEAT_INTERVAL
    end
  end

  # Defines how often Resque checks for dead workers.
  attr_writer :prune_interval
  def prune_interval
    if defined? @prune_interval
      @prune_interval
    else
      DEFAULT_PRUNE_INTERVAL
    end
  end

  # By default, jobs are pushed to the back of the queue and popped from
  # the front, resulting in "first in, first out" (FIFO) execution order.
  # Set to true to push jobs to the front of the queue instead, resulting
  # in "last in, first out" (LIFO) execution order.
  attr_writer :enqueue_front
  def enqueue_front
    if defined? @enqueue_front
      @enqueue_front
    else
      @enqueue_front = false
    end
  end

  # The `before_first_fork` hook will be run in the **parent** process
  # only once, before forking to run the first job. Be careful- any
  # changes you make will be permanent for the lifespan of the
  # worker.
  #
  # Call with a block to register a hook.
  # Call with no arguments to return all registered hooks.
  def before_first_fork(&block)
    block ? register_hook(:before_first_fork, block) : hooks(:before_first_fork)
  end

  # Register a before_first_fork proc.
  def before_first_fork=(block)
    register_hook(:before_first_fork, block)
  end

  # The `before_fork` hook will be run in the **parent** process
  # before every job, so be careful- any changes you make will be
  # permanent for the lifespan of the worker.
  #
  # Call with a block to register a hook.
  # Call with no arguments to return all registered hooks.
  def before_fork(&block)
    block ? register_hook(:before_fork, block) : hooks(:before_fork)
  end

  # Register a before_fork proc.
  def before_fork=(block)
    register_hook(:before_fork, block)
  end

  # The `after_fork` hook will be run in the child process and is passed
  # the current job. Any changes you make, therefore, will only live as
  # long as the job currently being processed.
  #
  # Call with a block to register a hook.
  # Call with no arguments to return all registered hooks.
  def after_fork(&block)
    block ? register_hook(:after_fork, block) : hooks(:after_fork)
  end

  # Register an after_fork proc.
  def after_fork=(block)
    register_hook(:after_fork, block)
  end

  # The `before_pause` hook will be run in the parent process before the
  # worker has paused processing (via #pause_processing or SIGUSR2).
  def before_pause(&block)
    block ? register_hook(:before_pause, block) : hooks(:before_pause)
  end

  # Register a before_pause proc.
  def before_pause=(block)
    register_hook(:before_pause, block)
  end

  # The `after_pause` hook will be run in the parent process after the
  # worker has paused (via SIGCONT).
  def after_pause(&block)
    block ? register_hook(:after_pause, block) : hooks(:after_pause)
  end

  # Register an after_pause proc.
  def after_pause=(block)
    register_hook(:after_pause, block)
  end

  # The `queue_empty` hook will be run in the **parent** process when
  # the worker finds no more jobs in the queue and becomes idle.
  #
  # Call with a block to register a hook.
  # Call with no arguments to return all registered hooks.
  def queue_empty(&block)
    block ? register_hook(:queue_empty, block) : hooks(:queue_empty)
  end

  # Register a queue_empty proc.
  def queue_empty=(block)
    register_hook(:queue_empty, block)
  end

  # The `worker_exit` hook will be run in the **parent** process
  # after the worker has existed (via SIGQUIT, SIGTERM, SIGINT, etc.).
  #
  # Call with a block to register a hook.
  # Call with no arguments to return all registered hooks.
  def worker_exit(&block)
    block ? register_hook(:worker_exit, block) : hooks(:worker_exit)
  end

  # Register a worker_exit proc.
  def worker_exit=(block)
    register_hook(:worker_exit, block)
  end

  def to_s
    "Resque Client connected to #{redis_id}"
  end

  attr_accessor :inline

  # If 'inline' is true Resque will call #perform method inline
  # without queuing it into Redis and without any Resque callbacks.
  # The 'inline' is false Resque jobs will be put in queue regularly.
  alias :inline? :inline

  #
  # queue manipulation
  #

  # Pushes a job onto a queue. Queue name should be a string and the
  # item should be any JSON-able Ruby object.
  #
  # Resque works generally expect the `item` to be a hash with the following
  # keys:
  #
  #   class - The String name of the job to run.
  #    args - An Array of arguments to pass the job. Usually passed
  #           via `class.to_class.perform(*args)`.
  #
  # Example
  #
  #   ResqueSqs.push('archive', :class => 'Archive', :args => [ 35, 'tar' ])
  #
  # Returns nothing
  def push(queue, item)
    @data_store.push_to_queue(queue,encode(item))
  end

  # Pops a job off a queue. Queue name should be a string.
  #
  # Returns a receipt_handle and Ruby object.
  def pop(queue)
    receipt_handle, body = @data_store.pop_from_queue(queue)
    [receipt_handle, decode(body)]
  end

  # Returns an integer representing the size of a queue.
  # Queue name should be a string.
  def size(queue)
    @data_store.queue_size(queue)
  end

  # Does the dirty work of fetching a range of items from a Redis list
  # and converting them into Ruby objects.
  def list_failed_queue_range(key, start = 0, count = 1)
    results = @data_store.list_failed_queue_range(key, start, count)
    if count == 1
      decode(results)
    else
      results.map { |result| decode(result) }
    end
  end

  # Returns an array of all known Resque queues as strings.
  def queues
    @data_store.queue_names
  end

  # Given a queue name, removes all elements from the queue.
  def purge_queue(queue)
    @data_store.purge_queue(queue)
  end

  def remove_from_queue(queue, receipt_handle)
    @data_store.remove_from_queue(queue, receipt_handle)
  end

  # Used internally to keep track of which queues we've created.
  # Don't call this directly.
  def watch_queue(queue)
    @data_store.watch_queue(queue)
  end


  #
  # job shortcuts
  #

  # This method can be used to conveniently add a job to a queue.
  # It assumes the class you're passing it is a real Ruby class (not
  # a string or reference) which either:
  #
  #   a) has a @queue ivar set
  #   b) responds to `queue`
  #
  # If either of those conditions are met, it will use the value obtained
  # from performing one of the above operations to determine the queue.
  #
  # If no queue can be inferred this method will raise a `ResqueSqs::NoQueueError`
  #
  # Returns true if the job was queued, nil if the job was rejected by a
  # before_enqueue hook.
  #
  # This method is considered part of the `stable` API.
  def enqueue(klass, *args)
    enqueue_to(queue_from_class(klass), klass, *args)
  end

  # Just like `enqueue` but allows you to specify the queue you want to
  # use. Runs hooks.
  #
  # `queue` should be the String name of the queue you're targeting.
  #
  # Returns true if the job was queued, nil if the job was rejected by a
  # before_enqueue hook.
  #
  # This method is considered part of the `stable` API.
  def enqueue_to(queue, klass, *args)
    # Perform before_enqueue hooks. Don't perform enqueue if any hook returns false
    before_hooks = Plugin.before_enqueue_hooks(klass).collect do |hook|
      klass.send(hook, *args)
    end
    return nil if before_hooks.any? { |result| result == false }

    Job.create(queue, klass, *args)

    Plugin.after_enqueue_hooks(klass).each do |hook|
      klass.send(hook, *args)
    end

    return true
  end

  # Given a class, try to extrapolate an appropriate queue based on a
  # class instance variable or `queue` method.
  def queue_from_class(klass)
    (klass.instance_variable_defined?(:@queue) && klass.instance_variable_get(:@queue)) ||
      (klass.respond_to?(:queue) and klass.queue)
  end

  # This method will return a `ResqueSqs::Job` object or a non-true value
  # depending on whether a job can be obtained. You should pass it the
  # precise name of a queue: case matters.
  #
  # This method is considered part of the `stable` API.
  def reserve(queue)
    Job.reserve(queue)
  end

  # Validates if the given klass could be a valid Resque job
  #
  # If no queue can be inferred this method will raise a `ResqueSqs::NoQueueError`
  #
  # If given klass is nil this method will raise a `ResqueSqs::NoClassError`
  def validate(klass, queue = nil)
    queue ||= queue_from_class(klass)

    if !queue
      raise NoQueueError.new("Jobs must be placed onto a queue. No queue could be inferred for class #{klass}")
    end

    if klass.to_s.empty?
      raise NoClassError.new("Jobs must be given a class.")
    end
  end

  #
  # worker shortcuts
  #

  # A shortcut to Worker.all
  def workers
    Worker.all
  end

  # A shortcut to Worker.working
  def working
    Worker.working
  end

  # A shortcut to unregister_worker
  # useful for command line tool
  def remove_worker(worker_id)
    worker = ResqueSqs::Worker.find(worker_id)
    worker.unregister_worker
  end

  #
  # stats
  #

  # Returns a hash, similar to redis-rb's #info, of interesting stats.
  def info
    return {
      :pending   => queue_sizes.inject(0) { |sum, (_queue_name, queue_size)| sum + queue_size },
      :processed => Stat[:processed],
      :queues    => queues.size,
      :workers   => workers.size.to_i,
      :working   => working.size,
      :failed    => @data_store.num_failed,
      :servers   => [redis_id],
      :environment  => ENV['RAILS_ENV'] || ENV['RACK_ENV'] || 'development'
    }
  end

  # Returns an array of all known Resque keys in Redis. Redis' KEYS operation
  # is O(N) for the keyspace, so be careful - this can be slow for big databases.
  def keys
    @data_store.all_resque_keys
  end

  # Returns a hash, mapping queue names to queue sizes
  def queue_sizes
    queue_names = queues
    sizes = []
    queue_names.each do |queue_name|
      sizes << ResqueSqs.size(queue_name)
    end

    Hash[queue_names.zip(sizes)]
  end

  private

  @hooks = Hash.new { |h, k| h[k] = [] }

  # Register a new proc as a hook. If the block is nil this is the
  # equivalent of removing all hooks of the given name.
  #
  # `name` is the hook that the block should be registered with.
  def register_hook(name, block)
    return clear_hooks(name) if block.nil?

    block = Array(block)
    @hooks[name].concat(block)
  end

  # Clear all hooks given a hook name.
  def clear_hooks(name)
    @hooks[name] = []
  end

  # Retrieve all hooks of a given name.
  def hooks(name)
    @hooks[name]
  end
end

# Log to STDOUT by default
ResqueSqs.logger           = MonoLogger.new(STDOUT)
ResqueSqs.logger.formatter = ResqueSqs::QuietFormatter.new
